src/LanguageServer/Impl/Sources/CodeActionSource.cs

```diff
@@ -0,0 +1,81 @@
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Python.Analysis;
+using Microsoft.Python.Analysis.Analyzer;
+using Microsoft.Python.Analysis.Diagnostics;
+using Microsoft.Python.Core;
+using Microsoft.Python.Core.Collections;
+using Microsoft.Python.LanguageServer.CodeActions;
+using Microsoft.Python.LanguageServer.Protocol;
+using Microsoft.Python.Parsing.Ast;
+using Range = Microsoft.Python.Core.Text.Range;
+
+namespace Microsoft.Python.LanguageServer.Sources {
+    internal sealed partial class CodeActionSource {
+        private static readonly ImmutableArray<ICodeActionProvider> _codeActionProviders =
+            ImmutableArray<ICodeActionProvider>.Create(MissingImportCodeActionProvider.Instance);
+
+        private readonly IServiceContainer _services;
+
+        public CodeActionSource(IServiceContainer services) {
+            _services = services;
+        }
+
+        public async Task<CodeAction[]> GetCodeActionsAsync(IDocumentAnalysis analysis, Range range, Diagnostic[] diagnostics, CancellationToken cancellationToken) {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var results = new List<CodeAction>();
+
+            // * NOTE * this always re-calculate whole document for all linters.
+            //          consider creating linter result cache somewhere (MRU?) and 
+            //          add error code as optional argument to run only that linter rather
+            //          than always running all linters.
+            //          also, current LintModule implementation always run all linters
+            //          even if the linter option is off to set references in the given
+            //          modules. this code path should use different code path to prevent that.
+            foreach (var diagnostic in GetMatchingDiagnostics(analysis, diagnostics, cancellationToken)) {
+                foreach (var codeActionProvider in _codeActionProviders) {
+                    if (codeActionProvider.FixableDiagnostics.Any(code => code == diagnostic.ErrorCode)) {
+                        results.AddRange(await codeActionProvider.GetCodeActionAsync(analysis, diagnostic, cancellationToken));
+                    }
+                }
+            }
+
+            return results.ToArray();
+        }
+
+        private IEnumerable<DiagnosticsEntry> GetMatchingDiagnostics(IDocumentAnalysis analysis, Diagnostic[] diagnostics, CancellationToken cancellationToken) {
+            var analyzer = _services.GetService<IPythonAnalyzer>();
+            foreach (var diagnostic in analysis.Diagnostics.Concat(analyzer.LintModule(analysis.Document))) {
```

Comment:
Analysis runs linter, do we need to re-run it here? All messages should be in the diagnostics

Reply:
we don't know whether the diagnostic is staled or not. running 1 particular linter seems not too bad?

Reply:
The analysis you use for linting is either current - and then the linting is current, or stale - then running linting on a stale analysis does not make sense.

Reply:
thank you for the explanation. changed it to use IDiagnosticService.